Add MELCloud integration

vilppuvuorinen opened the pull request:
## Description:

*This stuff is not supported nor in any way affiliated with Mitsubishi Electric.*

Provides climate and sensor platforms for Mitsubishi Electric heat pumps connected to MELCloud service. Multiple platforms on one go is not the best option, but it does not make sense to remove them and commit them later either. This thing started as a `custom_component`.

Email and access token are stored to the ConfigEntry. The token can be updated by adding the integration again with the same email address. The config flow is aborted and the update is performed on the background. This is a bit unorthodox, but at least I like it better than storing the plain-text password.

## Additional information

Link to documentation pull request: home-assistant/home-assistant.io#11930

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Commit:
Add MELCloud integration

* Provides a climate and sensor platforms. Multiple platforms on one go
is not the best option, but it does not make sense to remove them and
commit them later either.

* Email and access token are stored to the ConfigEntry. The token can be
updated by adding the integration again with the same email address. The
config flow is aborted and the update is performed on the background.

Commit:
Run isort

Commit:
Fix pylint errors

Commit:
Run black

Commit:
Increase coverage

Commit:
Update pymelcloud dependency

vilppuvuorinen commented:
@frenck Do I need to provide docs PR before any review is done? I can do the docs PR, but I don't want to commit time to doing that without knowing if this is something that can/will be merged.

frenck commented:
Docs are required before being able to merge. I've added it to ensure we do not merge it accidentally without documentation.

Commit:
Add HVAC_MODE_OFF emulation

Commit:
Remove print

rudizl commented:
Excellent work, fingers crossed!

gdude2002 commented:
I attempted to test this component on my own HA install, but it appears to assume that MELCloud is only used for HVAC/AC systems. When tested with my system (simply an air-to-water system), I noted that the component just presumed that I had an HVAC system. It also doesn't show any zones, and only the first sensor (the power one) reports anything.

![image](https://user-images.githubusercontent.com/204153/73214622-784e3c80-414a-11ea-927e-fa2755fd85ff.png)

Is this out of scope of this PR, or is this unintentional?

vilppuvuorinen commented:
I think the air-to-water systems are out of scope, but the lack of filtering is not. I also noticed I'm not filtering out empty model numbers for that device info like I though I did.

I have some read-only stuff for air-to-water systems [here](/vilppuvuorinen/hass-melcloud/tree/ecodan-support), but it has just temperatures for now. I would need help with that because I'm doing this with a blackbox or record/replay type approach (just in case) and therefore I don't have access to the source code of the original application. Help would be appreaciated. Right now it would be awesome to have captures of `/Device/Get` response body and bodies of individual parameter updates requests (`SetAta`) done at least 1 minute apart. That data is awful to anonymize. You can find me on home-assistant Discord.

rytilahti commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+            return self.async_abort(reason="cannot_connect")
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected error creating device")
+            return self.async_abort(reason="unknown")
```
Do not catch the bare exception, but let it bubble up.

vilppuvuorinen replied:
Done

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,158 @@
+    def __init__(self, device: Device):
+        """Construct a device wrapper."""
+        self.device = device
+        self.name = device.name
```
I/O is limited to `device.update` and `device.set` functions.

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,158 @@
+                conf_update_interval=timedelta(minutes=5),
+                device_set_debounce=timedelta(milliseconds=500),
+            )
+            devices = await client.get_devices()
```
I need to update the backend lib to fix the device aliasing issue mentioned above. This will change a bit there. The new implementation will look the same and it really does I/O in the timeout.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,200 @@
+    def hvac_modes(self) -> List[str]:
+        """Return the list of available hvac operation modes."""
+        return [HVAC_MODE_OFF] + list(
+            map(HVAC_MODE_LOOKUP.get, self._api.device.operation_modes())
```
No I/O. I'll switch the signature to property access with the backend lib update.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,200 @@
+        speed = self._api.device.fan_speed
+        if speed is None:
+            return None
+        return speed.replace("-", " ")
```
There's a supposedly dynamic amount of fan speeds available in these devices. The backend lib doing this `speed-%d` mapping for them and this is just to turn that "constant-y" string to "human readable" variant. What I'm trying to say is that I think the replace belongs here, but it just sucks.

Would it be better to use the `FAN_*` fan mode constants here instead of sequentially numbered speeds? I should be possible at least, but I don't know if it has any benefits.

rytilahti replied:
Sorry, I don't know about the details, but maybe using those constants will help when it comes to consistency (think: usage via voice controls)?

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,200 @@
+    @property
+    def fan_modes(self) -> Optional[List[str]]:
+        """Return the list of available fan modes."""
+        speeds = self._api.device.fan_speeds()
```
Nope.

Commit:
Update pymelcloud to enable device type filtering

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,158 @@
+            devices = await client.get_devices()
+    except asyncio.TimeoutError:
+        _LOGGER.debug("Connection timed out")
+        raise ConfigEntryNotReady
```
It looks like I can just collapse these blocks and do PEP 3134 exception chaining.

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,158 @@
+        raise ConfigEntryNotReady
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.error("Unexpected error when initializing client")
+        return None
```
Will be fixed after the blocks are collapsed.

Commit:
Collapse except blocks and chain ClientNotReadys

Commit:
Add preliminary documentation URL

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,200 @@
+
+    async def async_set_fan_mode(self, fan_mode: str) -> None:
+        """Set new target fan mode."""
+        await self._api.device.set({"fan_speed": fan_mode.replace(" ", "-")})
```
`FAN_*` built-in mode mappings could work here too.

MuppetOwl commented:
Swing modes (vertical vane adjusting) is missing Auto, Top, MiddleTop, Middle, MiddleBottom, Bottom and Swing
Also Heat/cool should maybe be renamed to Auto? Like it was in o0Zz version and is also what its called in Melcloud.

rytilahti commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,150 @@
+        unit_infos = self.device.units
+        if unit_infos is not None:
+            _device_info["model"] = ", ".join(
+                list(set(map(lambda x: x["model"], unit_infos)))
```
join accepts set iterable too, but I think using a list comprehension here would make it nicer to read:
```models = ", ".join(set([x["model"] for x  in unit_infos]))```

vilppuvuorinen replied:
Java Stream API is such a joy to use that I just can't learn to use list comprehension. All I can think is 'I wanna map these'. This really made it easy to filter out empty model names from the `unit_infos` too. Thanks!

rytilahti commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,205 @@
+
+from .const import DOMAIN, HVAC_MODE_LOOKUP, HVAC_MODE_REVERSE_LOOKUP, TEMP_UNIT_LOOKUP
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
```
Maybe `f"{DOMAIN}.{{}}"`?

vilppuvuorinen replied:
Fixed

rytilahti commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,205 @@
+        if name is None:
+            name = device.name
+
+        self._name = "{} {}".format(name, "HVAC")
```
This could also be an f-string.

vilppuvuorinen replied:
I just dropped the whole `"HVAC"` part.

rytilahti commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,205 @@
+    @property
+    def hvac_mode(self) -> str:
+        """Return hvac operation ie. heat, cool mode."""
+        mode = self._api.device.operation_mode
```
You access `_api.device` in several places, maybe it's worth to store it in `self._device`?

vilppuvuorinen replied:
Fixed.

rytilahti commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,205 @@
+        speeds = self._api.device.fan_speeds
+        if speeds is None:
+            return None
+        return list(map(lambda x: x.replace("-", " "), speeds))
```
Prefer list comprehension over map & list.

vilppuvuorinen replied:
I ended up just changing the format in the backend library and now this stuff is redundant. Fixed-ish.

vilppuvuorinen commented:
@MuppetOwl I'll have to backtrack on my comments on the forum. `HVAC_MODE_AUTO` is described as "The device is set to a schedule, learned behavior, AI." and that's definitely not the case here. I'll stick with `HVAC_MODE_HEAT_COOL`.

Commit:
Use list comp for creating model info

Filters out empty model names form units.

Commit:
f-string galore

Dropped 'HVAC' from AtaDevice name template.

Commit:
Delegate fan mode mapping to pymelcloud

Commit:
Fix type annotation

Commit:
Access AtaDevice through self._device

Commit:
Prefer list comprehension

rytilahti commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,150 @@
+async def async_setup(hass: HomeAssistantType, config: ConfigEntry):
+    """Establish connection with MELCloud."""
+    if DOMAIN not in config:
+        return True
```
When would this happen?

vilppuvuorinen replied:
Removed. Copy/paste was strong with these lines. I just happened to pick one of the rare integrations using this check.

vilppuvuorinen replied:
I was a little hasty with this change. This check here seems to be required if you want to allow an integration to be configured both through `configuration.yaml` and `config_flow`. I'd rather yank the whole `configuration.yaml` but I'm not sure if that makes sense.

rytilahti commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,150 @@
+    conf = entry.data
+    mel_api = await mel_api_setup(hass, conf[CONF_TOKEN])
+    if not mel_api:
+        return False
```
Should this rather raise an exception to inform the user about the failure?

vilppuvuorinen replied:
This line is unreachable. `mel_api_setup` (renamed to `mel_devices_setup`) already throws an exception on the single error path the function has.

rytilahti commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,150 @@
+        unit_infos = self.device.units
+        if unit_infos is not None:
+            _device_info["model"] = ", ".join(
+                [x["model"] for x in unit_infos if len(x["model"]) > 0]
```
`if x["model"]` is enough to check for noneness.

vilppuvuorinen replied:
The check is for empty strings. `device.units` can have valid entries with an empty string as the `model`.

rytilahti replied:
Empty strings are considered "false" (I assume because its length is 0): https://docs.python.org/3/library/stdtypes.html#truth-value-testing

vilppuvuorinen replied:
Of course. The term 'noneness' threw me off here. Javascript folks call these 'truthy' and 'falsy' and that's like _exactly_ what they are. It's weird they are not used with Python. I'll fix this.

rytilahti replied:
Sorry, my bad! The word choice in my original comment was absolutely incorrect and misleading.

rytilahti commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,200 @@
+        [
+            AtaDeviceClimate(mel_device)
+            for mel_device in mel_devices
+            if isinstance(mel_device.device, AtaDevice)
```
Using isinstance is generally a bad sign, as it does not allow easy testing (mocking is hard).

vilppuvuorinen replied:
A quick Google search on `isinstance` seems to agree.

I did a new release of the `pymelcloud` library that switches the return value from `get_devices` to a dict that's laid out like this: `{DeviceType: List[Device]}`. There's no real use case for having the devices in a single list so this seems natural enough. It also skips the filter step completely as an added bonus.

rytilahti reviewed (commented):
This looks fine to me, I left a couple of comments on very minor issues, maybe someone else can also take a quick look before getting this merged?

Commit:
Update pymelcloud to leverage device type grouping

The updated backend lib returns devices in a dict grouped by the device
type. The devices do not necessarily need to be in a single list and
this way isinstance is not required to extract devices by type.

Commit:
Remove DOMAIN presence check

This does not seem to make much sense after all.

Commit:
Fix async_setup_entry

Entry setup used half-baked naming from few experimentations back.
The naming conventiens were unified to match the platforms.

A redundant noneness check was also removed after evaluating the
possible return values from the backend lib.

rytilahti reviewed (approved):
Looks good to me, but considering the PRs merged by myself often get commented after the merge, I hope someone else will also take a quick look before we proceed with merging :-) (ping @MartinHjelmare ;-))

Commit:
Simplify empty model name check

Commit:
Improve config validation

* Use config_validation strings.

* Add CONF_EMAIL to config schema. The value is not strictly required
when configuring through configuration.yaml, but having it there makes
things more consistent.

* Use dict[key] to access required properties.

* Add DOMAIN in config check back to async_setup. This is required if an
integration is configured throught config_flow.

Commit:
Remove unused manifest properties

Commit:
Remove redundant ClimateDevice property override

vilppuvuorinen commented on homeassistant/components/melcloud/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "requirements": ["pymelcloud==2.0.0"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,148 @@
+
+CONF_LANGUAGE = "language"
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema({vol.Required(CONF_TOKEN): str})}, extra=vol.ALLOW_EXTRA,
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,148 @@
+async def async_setup(hass: HomeAssistantType, config: ConfigEntry):
+    """Establish connection with MELCloud."""
+    email = config[DOMAIN].get(CONF_EMAIL)
+    token = config[DOMAIN].get(CONF_TOKEN)
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,196 @@
+
+        False if entity pushes its state to HA.
+        """
+        return True
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,148 @@
+
+CONF_LANGUAGE = "language"
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema({vol.Required(CONF_TOKEN): str})}, extra=vol.ALLOW_EXTRA,
```
`CONF_EMAIL` is not really required to make things work through `configuration.yaml`. However, it makes sense to include it anyways just for the sake of consistency. I've added it to the schema.

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+
+@pytest.fixture
+def mock_request_info():
+    """Mock RequestInfo to create ClientResposenErrors."""
```
ClientResponseErrors

vilppuvuorinen replied:
Fixed.

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+)
+async def test_form_errors(hass, mock_login, mock_get_devices, error, reason):
+    """Test we handle cannot connect error."""
+    mock_login.return_value = mock_coro(exception=error)
```
Set a side_effect instead.

vilppuvuorinen replied:
Done

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
We can pass the user input as the `data` keyword argument if we don't need to test the initial form result again.

vilppuvuorinen replied:
Of course. I wish I had paid enough attention before writing all the tests with the form phase. Fixed.

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+async def test_failed_import_form(hass, mock_login, mock_get_devices):
+    """Test we get the form if imported without token."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data={},
```
This should never happen.

vilppuvuorinen replied:
Removed

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+async def test_token_refresh(hass, mock_login, mock_get_devices):
+    """Re-configuration with existing email should refresh token."""
+    await hass.config_entries.async_add(
+        config_entries.ConfigEntry(
```
Use `MockConfigEntry` imported from tests/common.py.

vilppuvuorinen replied:
Neat! Switched to the mock version.

MartinHjelmare reviewed (changes requested):
Don't we want to set up a way to refresh the token automatically?

vilppuvuorinen commented:
@MartinHjelmare Regarding the token update. I don't think it's possible to setup automatic token update without storing the password. There's a OAuth provider attached to the service, but that's aimed at Alexa integration and such. The response from customer service was hostile enough so that's definitely off the table. 

I was hoping I could avoid storing the password. Should I just bite the bullet and store the password and implement token refresh? The tokens seem to be extremely long lived and therefore it is also difficult to assess how the expiry logic would work.

Regarding imports: Should I use relative imports for importing things from the `homeassistant/components/melcloud` directory? 

I'll see about the other comments later today.

MartinHjelmare commented:
Ok, let's keep the token handling like it is for now.

We have helpers for implementing an oauth flow in the config flow if you want to try that in the future. Maybe you've already read that?

https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html#configuration-via-oauth2

Yes, intra-package imports should be relative.

Commit:
Add __init__.py to coverage exclusion

Commit:
Use CONF_USERNAME instead of CONF_EMAIL

Commit:
Use asyncio.gather instead of asyncio.wait

Commit:
Misc fixes

* any -> Any

* Better names for dict iterations

* Proper dict access with mandatory/known keys

* Remove unused 'name' argument

* Remove unnecessary platform info from unique_ids

* Remove redundant methods from climate platform

* Remove redundant default value from dict get

* Update ConfigFlow sub-classing

* Define sensors in a dict instead of a list

Commit:
Use _abort_if_unique_id_configured to update token

Commit:
Fix them tests

vilppuvuorinen commented on .coveragerc:
```diff
@@ -413,6 +413,8 @@ omit =
     homeassistant/components/mcp23017/*
     homeassistant/components/media_extractor/*
     homeassistant/components/mediaroom/media_player.py
+    homeassistant/components/melcloud/climate.py
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,157 @@
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {vol.Required(CONF_EMAIL): cv.string, vol.Required(CONF_TOKEN): cv.string}
```
Done. I'm still referring to this value email in from and messages. That's what the user is supplying to the field anyways.

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,157 @@
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    await asyncio.wait(
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,157 @@
+            _LOGGER.warning("Connection failed for %s", self.name)
+            self._available = False
+
+    async def async_set(self, properties: Dict[str, any]):
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,157 @@
+        raise ConfigEntryNotReady() from ex
+
+    wrapped_devices = {}
+    for k, value in devices.items():
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+):
+    """Set up MelCloud device climate based on config_entry."""
+    mel_devices = hass.data[DOMAIN].get(entry.entry_id)
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+
+from .const import DOMAIN, HVAC_MODE_LOOKUP, HVAC_MODE_REVERSE_LOOKUP, TEMP_UNIT_LOOKUP
+
+ENTITY_ID_FORMAT = f"{DOMAIN}.{{}}"
```
Removed.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+class AtaDeviceClimate(ClimateDevice):
+    """Air-to-Air climate device."""
+
+    def __init__(self, device: MelCloudDevice, name=None):
```
I though it looked weird, but I didn't have the guts to just scrap it. Removed.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return f"{self._device.serial}-{self._device.mac}-climate"
```
Suffix removed.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+        return self._api.device_info
+
+    @property
+    def state(self) -> str:
```
Removed. I do have a follow up PR in mind where `state` would not necessarily match `hvac_mode`. `hvac_modes` would be a subset of possible states. Does this sound like something that would work?

MartinHjelmare replied:
No, the climate entity state must be one of the hvac modes. We have to follow the entity model defined in our architecture.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+        return self.hvac_mode
+
+    @property
+    def precision(self) -> float:
```
Removed. I think I can find the real measurement precision from the system. The default will do in the meantime.

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+            await self._device.set({"power": False})
+            return
+
+        operation_mode = HVAC_MODE_REVERSE_LOOKUP.get(hvac_mode, None)
```
Removed.

vilppuvuorinen commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@config_entries.HANDLERS.register("melcloud")
+class FlowHandler(config_entries.ConfigFlow):
```
Done

vilppuvuorinen commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    async def _create_entry(self, email: str, token: str):
+        """Register new entry."""
+        for entry in self._async_current_entries():
+            if entry.data.get(CONF_EMAIL, entry.title) == email:
```
I tried asking around for a similar feature a while back but I guess it was still either in the making or just really fresh. I had to rebase on top of `dev` to get this in. Neat!

vilppuvuorinen commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+    async def async_step_import(self, user_input):
+        """Import a config entry."""
+        email = user_input.get(CONF_EMAIL)
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        """Import a config entry."""
+        email = user_input.get(CONF_EMAIL)
+        token = user_input.get(CONF_TOKEN)
+        if not token:
```
Most likely never. Removed.

vilppuvuorinen commented on homeassistant/components/melcloud/sensor.py:
```diff
@@ -0,0 +1,102 @@
+ATTR_DEVICE_CLASS = "device_class"
+ATTR_VALUE_FN = "value_fn"
+
+SENSORS = [
```
Done.

vilppuvuorinen commented on homeassistant/components/melcloud/sensor.py:
```diff
@@ -0,0 +1,102 @@
+        self._api = device
+        if name is None:
+            self._name_slug = device.name
+        else:
```
Removed.

vilppuvuorinen commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+import asyncio
+
+from aiohttp import ClientError, ClientResponseError
+from asynctest import patch as async_patch
```
Done.

vilppuvuorinen commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+from homeassistant import config_entries
+from homeassistant.components.melcloud.const import DOMAIN
+
+from tests.common import mock_coro
```
Done.

vilppuvuorinen commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+def mock_login():
+    """Mock pymelcloud login."""
+    with async_patch("pymelcloud.login") as mock:
+        mock.return_value = mock_coro("test-token")
```
Done.

vilppuvuorinen commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,204 @@
+def mock_get_devices():
+    """Mock pymelcloud get_devices."""
+    with async_patch("pymelcloud.get_devices") as mock:
+        mock.return_value = mock_coro(
```
Done.

MartinHjelmare commented on homeassistant/components/melcloud/__init__.py:
```diff
@@ -0,0 +1,160 @@
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    await asyncio.gather(
+        [
```
```suggestion
        *[
```

MartinHjelmare commented on homeassistant/components/melcloud/sensor.py:
```diff
@@ -0,0 +1,96 @@
+ATTR_VALUE_FN = "value_fn"
+
+SENSORS = {
+    "Room Temperature": {
```
I'd separate the sensor type from the measurement name.
```py
SENSORS = {
    "room_temperature": {
        ATTR_MEASUREMENT: "Room Temperature",
        ...
    ...
}
```

MartinHjelmare commented on homeassistant/components/melcloud/sensor.py:
```diff
@@ -0,0 +1,96 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        normalized = self._measurement.lower().replace(" ", "_")
```
Then we can remove this.

MartinHjelmare commented on tests/components/melcloud/test_config_flow.py:
```diff
@@ -0,0 +1,172 @@
+
+async def test_token_refresh(hass, mock_login, mock_get_devices):
+    """Re-configuration with existing username should refresh token."""
+    await hass.config_entries.async_add(
```
Don't use this. Use `MockConfigEntry.add_to_hass`.
https://github.com/home-assistant/home-assistant/blob/41f3fb291afe6596f276400e8c692e815172f40d/tests/common.py#L692

Commit:
Remove current state guards

vilppuvuorinen commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -0,0 +1,188 @@
+
+    async def async_turn_off(self) -> None:
+        """Turn the entity off."""
+        if self._device.power:
```
There's no point in guarding here. The device don't care. Removed the guards.

Commit:
Fix that gather call

Commit:
Implement sensor definitions without str manipulation

Commit:
Use relative intra-package imports

vilppuvuorinen commented:
That test failure does not make sense to me. As if pylint didn't see the usage at `homeassistant/components/melcloud/config_flow.py:19`.

MartinHjelmare commented on homeassistant/components/melcloud/config_flow.py:
```diff
@@ -0,0 +1,84 @@
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_TOKEN, CONF_USERNAME
+
+from .const import DOMAIN
```
This is a known bug in pylint. We'll disable it for now.
```suggestion
from .const import DOMAIN  # pylint: disable=unused-import
```

Commit:
Update homeassistant/components/melcloud/config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Nice!

vilppuvuorinen commented:
Should I squash the commits now?

MartinHjelmare commented:
No, please don't do that. We'll squash when merging.

springstan reviewed (approved):
LGTM 🎉

rytilahti commented:
Considering we have already three approvals, let's get this merged. Thanks @vilppuvuorinen! :tada:
